### PR TITLE
fix(36416): ERR <semantic> TypeScript Server Error

### LINKED ISSUE #36416: ERR <semantic> TypeScript Server Error (3.7.5)
*TS Template added by @mjbvz*

**TypeScript Version**: 3.8.0-dev.20200124 and 3.7.5

---

* open a `.ts` file with the following content:

```
import * as http from ''
```

* place cursor before `''`
* observe:
```
 ERR <semantic> TypeScript Server Error (3.7.5)
Cannot read property 'parent' of undefined
TypeError: Cannot read property 'parent' of undefined
    at Object.deleteDeclaration (tsserver.js:120255:56)
    at _loop_9 (tsserver.js:119840:47)
    at ChangeTracker.finishDeleteDeclarations (tsserver.js:119847:21)
    at ChangeTracker.getChanges (tsserver.js:119867:22)
    at Function.ChangeTracker.with (tsserver.js:119391:32)
    at Object.getCodeActions (tsserver.js:124870:68)
    at tsserver.js:120431:121
    at Object.flatMap (tsserver.js:468:25)
    at Object.getFixes (tsserver.js:120431:23)
    at tsserver.js:131114:35
    at Object.flatMap (tsserver.js:468:25)
    at Proxy.getCodeFixesAtPosition (tsserver.js:131112:23)
    at IOSession.Session.getCodeFixes (tsserver.js:140767:64)
    at Session.handlers.ts.createMapFromTemplate._a.<computed> (tsserver.js:139552:61)
    at tsserver.js:140960:88
    at IOSession.Session.executeWithRequestId (tsserver.js:140951:28)
    at IOSession.Session.executeCommand (tsserver.js:140960:33)
    at IOSession.Session.onMessage (tsserver.js:140983:35)
    at Interface.<anonymous> (tsserver.js:142298:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17): Error: <semantic> TypeScript Server Error (3.7.5)
Cannot read property 'parent' of undefined
TypeError: Cannot read property 'parent' of undefined
    at Object.deleteDeclaration (tsserver.js:120255:56)
    at _loop_9 (tsserver.js:119840:47)
    at ChangeTracker.finishDeleteDeclarations (tsserver.js:119847:21)
    at ChangeTracker.getChanges (tsserver.js:119867:22)
    at Function.ChangeTracker.with (tsserver.js:119391:32)
    at Object.getCodeActions (tsserver.js:124870:68)
    at tsserver.js:120431:121
    at Object.flatMap (tsserver.js:468:25)
    at Object.getFixes (tsserver.js:120431:23)
    at tsserver.js:131114:35
    at Object.flatMap (tsserver.js:468:25)
    at Proxy.getCodeFixesAtPosition (tsserver.js:131112:23)
    at IOSession.Session.getCodeFixes (tsserver.js:140767:64)
    at Session.handlers.ts.createMapFromTemplate._a.<computed> (tsserver.js:139552:61)
    at tsserver.js:140960:88
    at IOSession.Session.executeWithRequestId (tsserver.js:140951:28)
    at IOSession.Session.executeCommand (tsserver.js:140960:33)
    at IOSession.Session.onMessage (tsserver.js:140983:35)
    at Interface.<anonymous> (tsserver.js:142298:27)
    at Interface.emit (events.js:203:13)
    at Interface._onLine (readline.js:316:10)
    at Interface._normalWrite (readline.js:461:12)
    at Socket.ondata (readline.js:172:10)
    at Socket.emit (events.js:203:13)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:276:11)
    at Socket.Readable.push (_stream_readable.js:210:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
	at Function.create (/Users/alex/src/vscode/extensions/typescript-language-features/out/tsServer/serverError.js:17:16)
	at ProcessBasedTsServer.dispatchResponse (/Users/alex/src/vscode/extensions/typescript-language-features/out/tsServer/server.js:141:66)
	at ProcessBasedTsServer.dispatchMessage (/Users/alex/src/vscode/extensions/typescript-language-features/out/tsServer/server.js:83:26)
	at /Users/alex/src/vscode/extensions/typescript-language-features/out/tsServer/server.js:56:41
	at Emitter.fire (/Users/alex/src/vscode/out/vs/base/common/event.js:459:38)
	at Reader.onLengthData (/Users/alex/src/vscode/extensions/typescript-language-features/out/utils/wireProtocol.js:111:30)
	at Socket.<anonymous> (/Users/alex/src/vscode/extensions/typescript-language-features/out/utils/wireProtocol.js:90:42)
	at Socket.emit (events.js:203:13)
	at addChunk (_stream_readable.js:295:12)
	at readableAddChunk (_stream_readable.js:276:11)
	at Socket.push (_stream_readable.js:210:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:166:17)
```

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #36416

### CODE REVIEW

File: src/services/textChanges.ts [hunk 1]
```diff
@@ -1238,9 +1238,10 @@ namespace ts.textChanges {
                 }
 
                 case SyntaxKind.ImportDeclaration:
+                    const isFirstImport = sourceFile.imports.length && node === first(sourceFile.imports).parent || node === find(sourceFile.statements, isImportDeclaration);
```
Comment (DanielRosenwasser) [inline src/services/textChanges.ts]: When would

```
sourceFile.imports.length && node === first(sourceFile.imports).parent
```

be true, but not the following?

```ts
node === find(sourceFile.statements, isImportDeclaration)
```

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-3.8

File: src/services/textChanges.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/textChanges.ts]: @DanielRosenwasser 

> When would sourceFile.imports.length && node === first(sourceFile.imports).parent


```ts
import { a } from "./b";
import { b } from "./c";
```
when each import statemnt has path

> be true, but not the following? node === find(sourceFile.statements, isImportDeclaration)

```ts
import a from "";
import { a } from "./b";
import { b } from "./c";
```

when there is an import statement with the empty path

**Comment by sandersn** [comment]
@sheetalkamat @DanielRosenwasser sounds like this should be merged into master too, right?

**Comment by DanielRosenwasser** [comment]
Probably

@typescript-bot pack this

File: src/services/textChanges.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/services/textChanges.ts]: It sounds like it won't work correctly when

```ts
import { a } from "./b";
// try to remove the following import
import a from "";
import { b } from "./c";
```

Does that still exhibit the same error?

Comment (a-tarasyuk) [reply src/services/textChanges.ts]: @DanielRosenwasser Sorry for the delay with the response, sometimes different time zones matter :). There is a test for a similar case only with leading comment 





Is it an answer to your question?

**Pull request merged**